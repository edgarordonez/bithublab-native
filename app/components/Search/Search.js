// @flow

import React from 'react';
import { StyleSheet, View, TextInput } from 'react-native';
import { Subject } from 'rxjs';
import { filter, debounceTime, distinctUntilChanged } from 'rxjs/operators';

type Props = {
  handleOnSearch: string => void,
};

const search$ = new Subject();
const handleOnChange = (value: string) => search$.next(value);

export const Search = ({ handleOnSearch }: Props) => {
  search$
    .pipe(
      debounceTime(150),
      filter(text => text.replace(/\s/g, '') !== ''),
      filter(text => text.length >= 2),
      debounceTime(425),
      distinctUntilChanged(),
    )
    .subscribe(handleOnSearch);

  return (
    <View style={style.container}>
      <TextInput
        style={style.input}
        placeholder="Search the magical world of OSS..."
        onChangeText={handleOnChange}
        selectionColor="#2684ff"
        underlineColorAndroid="transparent"
      />
    </View>
  );
};

const style = StyleSheet.create({
  container: {
    margin: 10,
    borderRadius: 5,
    elevation: 1,
  },
  input: {
    padding: 10,
    borderRadius: 5,
  },
});
