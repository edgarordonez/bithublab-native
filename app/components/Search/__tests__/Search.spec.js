// @noflow

import React from 'react';
import TestRenderer from 'react-test-renderer';
import ShallowRenderer from 'react-test-renderer/shallow';

import { Search } from '../Search';

const handleOnSearch = jest.fn();

const Props = {
  handleOnSearch,
};

describe('Search', () => {
  it('Renders without crashing', () => {
    const rendered = TestRenderer.create(<Search {...Props} />);
    expect(rendered).toMatchSnapshot();
  });

  it('Should call "handleOnSearch" when the event is triggered', () => {
    const value = 'D3-Stencil';
    jest.useFakeTimers();

    const renderer = new ShallowRenderer();
    renderer.render(<Search {...Props} />);

    const rendered = renderer.getRenderOutput();
    rendered.props.children.props.onChangeText(value);

    jest.runAllTimers();

    expect(handleOnSearch).toBeCalledWith(value);
  });
});
