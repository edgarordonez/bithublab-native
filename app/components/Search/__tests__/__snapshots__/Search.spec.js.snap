// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search Renders without crashing 1`] = `
<View
  style={
    Object {
      "borderRadius": 5,
      "elevation": 1,
      "margin": 10,
    }
  }
>
  <TextInput
    allowFontScaling={true}
    onChangeText={[Function]}
    placeholder="Search the magical world of OSS..."
    selectionColor="#2684ff"
    style={
      Object {
        "borderRadius": 5,
        "padding": 10,
      }
    }
    underlineColorAndroid="transparent"
  />
</View>
`;
